src/plugins/ui_feature/find_and_replace.ts

```diff
@@ -138,35 +152,116 @@ export class FindAndReplacePlugin extends UIPlugin {
     }
     this.currentSearchRegex = RegExp(searchValue, flags);
   }
+
   /**
    * Find matches using the current regex
    */
   private findMatches() {
-    const sheetId = this.getters.getActiveSheetId();
-    const cells = this.getters.getCells(sheetId);
     const matches: SearchMatch[] = [];
+    let sheets: UID[];
+    const specificRange = this.searchOptions.specificRange;
+    const activeSheetId = this.getters.getActiveSheetId();
+    if (!this.toSearch) {
+      return matches;
+    }
 
-    if (this.toSearch) {
-      for (const cell of Object.values(cells)) {
-        const { col, row } = this.getters.getCellPosition(cell.id);
-        if (
-          cell &&
-          this.currentSearchRegex &&
-          this.currentSearchRegex.test(this.getSearchableString({ sheetId, col, row }))
-        ) {
-          const match: SearchMatch = { col, row, selected: false };
-          matches.push(match);
+    switch (this.searchOptions.searchScope) {
+      case "allSheets":
+        sheets = this.getters.getSheetIds();
+        const activeSheetIndex = sheets.findIndex((id) => id === activeSheetId);
+        sheets = [
+          sheets[activeSheetIndex],
+          ...sheets.slice(activeSheetIndex + 1),
+          ...sheets.slice(0, activeSheetIndex),
+        ];
+        break;
+      case "thisSheet":
+        sheets = [activeSheetId];
+        break;
+      case "specificRange":
+        let rangeSheet = activeSheetId;
+        const specificRange = this.searchOptions.specificRange;
+        if (!specificRange) {
+          return [];
         }
-      }
+        if (specificRange.includes("!")) {
+          [rangeSheet] = specificRange.split("!");
+          rangeSheet = this.getters.getSheetIdByName(rangeSheet.replace(/'/g, "")) as UID;
+        }
+        sheets = [rangeSheet];
+        break;
     }
-    return matches.sort(this.sortByRowThenColumn);
+
+    for (const sheetId of sheets) {
+      matches.push(...this.findMatchesInSheet(sheetId, specificRange));
+    }
+    const sortedMatches = matches.sort((a, b) =>
+      this.sortBySheetinitialStateIndexThenRowThenColumn(a, b)
+    );
+
+    const activeSheetIndex = sortedMatches.findIndex((match) => match.sheetId === activeSheetId);
+    this.selectedMatchOffset = activeSheetIndex;
+
+    return sortedMatches;
   }
 
-  private sortByRowThenColumn(a: Position, b: Position) {
-    if (a.row === b.row) {
-      return a.col - b.col;
+  private findMatchesInSheet(sheetId: string, specificRange?: string) {
+    const matches: SearchMatch[] = [];
+    const rangeData = specificRange
+      ? this.getters.getRangeDataFromXc(sheetId, specificRange)._zone
+      : this.getters.getSheetZone(sheetId);
+
+    const boundedZone: Zone = {
+      left: rangeData.left,
+      top: rangeData.top,
+      right: rangeData.right ?? this.getters.getNumberCols(sheetId),
+      bottom: rangeData.bottom ?? this.getters.getNumberRows(sheetId),
+    };
```

Comment:
Previously, the cells were obtained using the "getCells" method, which iterated over the entire sheet. However, in cases where a specific range is required, this method is unnecessary. In order to address this issue, the range data is determined and the zone is retrieved if a specific range is present. However, it should be noted that the "getRangeDataFromXc" method may return an unbounded zone, thus requiring the zone to be bounded.

If the "getEvaluatedCellsInZone" method is used instead of "getCells", the cell ID is not received automatically. Therefore, it is necessary to manually add the columns and rows to determine the range.

Please let me know if there is any other possible way to deal with this situation.

Reply:
https://github.com/odoo/o-spreadsheet/pull/2330#discussion_r1169737983

Reply:
Hence the comment from @hokolomopo that you should NOT use  `getRangeDataFromXc` but rather use `getRangeFromSheetXc` .This will yield a `Range` `which has a public attribute `range.zone: Zone` which, as you can see in the definition of its class, will be a `Zone`and not an ÙnboundedZone`.  So the general comment in the last review of Adrien stands, please address every remark beforehand - we're not making progress here ...